### DISCUSSION TITLE
Seeking Assistance with Index-Based Rotation Strategy Implementation

### DISCUSSION THREAD

**Discussion by McHill007** [edtechre/pybroker#71] [Q&A] [answered]
I am currently working on a trading strategy that relies on a proprietary index and a self-developed indicator. I have successfully implemented a rotation strategy that buys a calculated set of stocks and rotates them.

However, I am encountering two issues:

- My strategy is based on the values of NASDAQ 100, S&P 500, DAX, and CAC 40. To conduct a backtest over multiple years (e.g., starting from the year 2000), I need a way to rotate stocks based on their index affiliation. Is there a possible solution for this?

- I am facing challenges when mixing stocks from different indices, such as NASDAQ 100 stocks with DAX stocks. This leads to an error: decimal.InvalidOperation: [<class 'decimal.InvalidOperation'>]. Is this due to different trading dates based on holidays etc..? Is there any solution to workaround. 

I must admit that I am not a Python expert, but I am trying to navigate through with my Perl knowledge and documentation. So far, I find the module really great, as it has immediately reflected my rotation strategy.

I would greatly appreciate any guidance or suggestions on how to address these issues. Thank you in advance for your help!

Code to reproduce the error.  'AIR.PA' seems to be the problem here, _verify_input fails since fill_price is nan
```
import pybroker as pyb
from pybroker import ExecContext, Strategy, StrategyConfig, YFinance
import numpy as np
from numba import njit


def cmma(bar_data, lookback):
    @njit  # Enable Numba JIT.
    def vec_cmma(values):
        # Initialize the result array.
        n = len(values)
        out = np.array([np.nan for _ in range(n)])

        # For all bars starting at lookback:
        for i in range(lookback, n):
            # Calculate the moving average for the lookback.
            ma = 0
            for j in range(i - lookback, i):
                ma += values[j]
            ma /= lookback
            # Subtract the moving average from value.
            out[i] = values[i] - ma
        return out

    # Calculate with close prices.
    return vec_cmma(bar_data.close)


cmma_20 = pyb.indicator('cmma_20', cmma, lookback=20)

config = StrategyConfig(max_long_positions=1, initial_cash=100_000)
pyb.param('target_size', 1 / config.max_long_positions)
pyb.param('rank_threshold', 5)


def rank(ctxs: dict[str, ExecContext]):
    scores = {
        symbol: ctx.indicator('cmma_20')[-1]
        for symbol, ctx in ctxs.items()
    }
    sorted_scores = sorted(
        scores.items(),
        key=lambda score: score[1],
        reverse=True
    )
    threshold = pyb.param('rank_threshold')
    top_scores = sorted_scores[:threshold]
    top_symbols = [score[0] for score in top_scores]
    pyb.param('top_symbols', top_symbols)


def rotate(ctx: ExecContext):
    if ctx.long_pos():
        if ctx.symbol not in pyb.param('top_symbols'):
            ctx.sell_all_shares()
    else:
        target_size = pyb.param('target_size')
        ctx.buy_shares = ctx.calc_target_shares(target_size)
        ctx.score = ctx.indicator('cmma_20')[-1]


strategy = Strategy(
    YFinance(),
    start_date='1/1/2018',
    end_date='1/1/2023',
    config=config
)
strategy.set_before_exec(rank)
strategy.add_execution(rotate, [
    'META',
    'AIR.PA',
    'RHM.DE'
], indicators=cmma_20)

result = strategy.backtest(warmup=20)

```

**Comment by edtechre** [comment, marked as answer]
Hi @McHill007,

Sorry for the late response.

> My strategy is based on the values of NASDAQ 100, S&P 500, DAX, and CAC 40. To conduct a backtest over multiple years (e.g., starting from the year 2000), I need a way to rotate stocks based on their index affiliation. Is there a possible solution for this?

You can use [pybroker.param](https://www.pybroker.com/en/latest/notebooks/FAQs.html#...-get-and-set-a-global-parameter?) to store index constituents and check for membership in your strategy e.g. 

`pybroker.param('qqq', ...)`

`if ctx.symbol in pybroker.param('qqq'):`


> I am facing challenges when mixing stocks from different indices, such as NASDAQ 100 stocks with DAX stocks. This leads to an error: decimal.InvalidOperation: [<class 'decimal.InvalidOperation'>]. Is this due to different trading dates based on holidays etc..? Is there any solution to workaround.

It looks like this is due to bad data being returned from YF as NaN. By default, the bar's close price is used as the fill price. Running this code:

```
yf = YFinance()
df = yf.query([
    'META',
    'AIR.PA',
    'RHM.DE'
], start_date='1/1/2018', end_date='1/1/2023')
df[df['close'].isna()]
```
Returns 68 rows with NaN close prices.

What you can do is query your data separately, drop all NaN values, and then use the resulting DataFrame as your Strategy's DataSource:

```
yf = YFinance()
df = yf.query([
    'META',
    'AIR.PA',
    'RHM.DE'
], start_date='1/1/2018', end_date='1/1/2023')
df = df.dropna()

strategy = Strategy(
    df,
    start_date='1/1/2018',
    end_date='1/1/2023',
    config=config
)

```
